apps/oracles/exsat-holdings/src/lib.rs

```diff
@@ -0,0 +1,368 @@
+    pub fn next(&self, prev: &AddrMappingsRows) -> Option<AddrMappings> {
+        if prev.more {
+            let mut res = self.clone();
+            res.lower_bound = Some(prev.next_key.clone());
+            Some(res)
+        } else {
+            None
+        }
+    }
+
+    pub async fn fetch_rows(&self) -> Result<AddrMappingsRows> {
+        let url = "https://rpc-us.exsat.network/v1/chain/get_table_rows";
+        let mut req = Request::builder();
+        req.method(Method::Post);
+        req.uri(url);
+        let data = serde_json::to_string(&self)?;
+        req.body(data.into_body());
+        let resp: Response = send(req).await?;
+        let body = resp.into_body();
+        let string = String::from_utf8(body)?;
+        let value: AddrMappingsRows = serde_json::from_str(&string)
+            .context(format!("Couldn't parse response from {url} properly"))?;
+        Ok(value)
+    }
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+pub struct AddrMappingData {
+    // pub id: u32,
+    // pub b_id: String,
+    // pub wallet_code: String,
+    pub btc_address: String,
+    // pub evm_address: String,
+    // pub last_bridge_time: String,
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+pub struct AddrMappingsRow {
+    pub data: AddrMappingData,
+    // pub payer: String,
+}
+
+#[derive(Debug, Default, Clone, Deserialize, Serialize)]
+pub struct AddrMappingsRows {
+    pub rows: Vec<AddrMappingsRow>,
+    pub more: bool,
+    pub next_key: String,
+}
+
+#[derive(Serialize, Deserialize, Debug)]
+pub struct BTCAddressValue {
+    pub final_balance: u128,
+    pub total_received: u128,
+    pub n_tx: u64,
+}
+type BTCAddress = String;
+type Dictionary = HashMap<BTCAddress, BTCAddressValue>;
+
+pub async fn fetch_balance(addresses: &Vec<BTCAddress>, store: &Store) -> Result<Dictionary> {
+    let base_url =
+        Url::parse("https://mempool.space/api/address/").expect("This should never happen");
+    let mut res = Dictionary::new();
+    for address in addresses {
+        if store.exists(address.as_str())? {
+            // We only store addresses that no longer will hold any funds
+            continue;
+        }
+
+        let url = base_url.join(address.as_str()).expect("msg");
+        let mut req = Request::builder();
+        req.method(Method::Get);
+        req.uri(url);
+        let resp: Response = send(req).await?;
+
+        let body = resp.into_body();
+        let string = String::from_utf8(body)?;
+        let value: BTCAddressStats = serde_json::from_str(&string)
+            .context(format!("Couldn't parse response from {base_url} properly"))?;
+
+        let key = value.address.as_str();
+        if value.should_skip() && !store.exists(key)? {
+            store.set_json(key, &value)?;
+        }
+
+        res.insert(value.address.clone(), value.pending());
+        // Sleep for one second, so we don't hit API rate limit
+        //thread::sleep(Duration::from_millis(1000));
```

Comment:
What is the status of this? Do we need it?

Reply:
I had several successful runs without this sleep, but I am not sure that will always be the case. Let's try it like this for now